In [1]:
import pandas as pd
import numpy as np
import folium
from sklearn.cluster import KMeans
import requests 
from pandas.io.json import json_normalize 
import matplotlib.cm as cm
import matplotlib.colors as colors

### Grab all the tables from the wikipedia website

In [2]:
tables = pd.read_html("https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M")

In [3]:
len(tables)

3

### The list contains 3 tables. We can use df.head() to find out the table we need.

In [4]:
print (tables[0].head())
print (tables[1].head())
print (tables[2].head())

  Postcode           Borough      Neighborhood
0      M1A      Not assigned      Not assigned
1      M2A      Not assigned      Not assigned
2      M3A        North York         Parkwoods
3      M4A        North York  Victoria Village
4      M5A  Downtown Toronto      Harbourfront
                                                  0   \
0                                                NaN   
1  NL NS PE NB QC ON MB SK AB BC NU/NT YT A B C E...   
2                                                 NL   
3                                                  A   

                                                  1   \
0                              Canadian postal codes   
1  NL NS PE NB QC ON MB SK AB BC NU/NT YT A B C E...   
2                                                 NS   
3                                                  B   

                                                  2    3    4    5    6    7   \
0                                                NaN  NaN  NaN  NaN  NaN  N

### The first one is the postal code table. Let's check the table info.

In [5]:
zipcode=tables[0]
print(zipcode.shape)
print(zipcode.describe())
zipcode.head(10)

(287, 3)
       Postcode       Borough  Neighborhood
count       287           287           287
unique      180            12           208
top         M9V  Not assigned  Not assigned
freq          8            77            78


Postcode           Borough      Neighborhood
0      M1A      Not assigned      Not assigned
1      M2A      Not assigned      Not assigned
2      M3A        North York         Parkwoods
3      M4A        North York  Victoria Village
4      M5A  Downtown Toronto      Harbourfront
5      M6A        North York  Lawrence Heights
6      M6A        North York    Lawrence Manor
7      M7A  Downtown Toronto      Queen's Park
8      M8A      Not assigned      Not assigned
9      M9A      Queen's Park      Not assigned

### Remove the cells without an assigned Borough. According to the describe(), there are 77 lines to be deleted.

In [6]:
zipcode= zipcode[zipcode['Borough']!='Not assigned']
zipcode.head(10)

Postcode           Borough      Neighborhood
2       M3A        North York         Parkwoods
3       M4A        North York  Victoria Village
4       M5A  Downtown Toronto      Harbourfront
5       M6A        North York  Lawrence Heights
6       M6A        North York    Lawrence Manor
7       M7A  Downtown Toronto      Queen's Park
9       M9A      Queen's Park      Not assigned
10      M1B       Scarborough             Rouge
11      M1B       Scarborough           Malvern
13      M3B        North York   Don Mills North

### Check the 'Not assigned' Neighborhood. Assigen the values.

In [7]:
zipcode.loc[zipcode.Neighborhood == 'Not assigned','Neighborhood']=zipcode.loc[zipcode.Neighborhood == 'Not assigned','Borough']

C:\Anaconda\lib\site-packages\pandas\core\indexing.py:190: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  self._setitem_with_indexer(indexer, value)
C:\Anaconda\lib\site-packages\ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  """Entry point for launching an IPython kernel.


In [8]:
zipcode.head(10)

Postcode           Borough      Neighborhood
2       M3A        North York         Parkwoods
3       M4A        North York  Victoria Village
4       M5A  Downtown Toronto      Harbourfront
5       M6A        North York  Lawrence Heights
6       M6A        North York    Lawrence Manor
7       M7A  Downtown Toronto      Queen's Park
9       M9A      Queen's Park      Queen's Park
10      M1B       Scarborough             Rouge
11      M1B       Scarborough           Malvern
13      M3B        North York   Don Mills North

### Group the Neighborhood column by the Postalcode and Borough. Separate them by comma.

In [9]:
def addon(df):
    return ','.join(df.values)

zip_grp=zipcode.groupby(['Postcode','Borough'])['Neighborhood'].apply(addon)

In [10]:
zip_grp.head(10)

Postcode  Borough    
M1B       Scarborough                                    Rouge,Malvern
M1C       Scarborough             Highland Creek,Rouge Hill,Port Union
M1E       Scarborough                  Guildwood,Morningside,West Hill
M1G       Scarborough                                           Woburn
M1H       Scarborough                                        Cedarbrae
M1J       Scarborough                              Scarborough Village
M1K       Scarborough        East Birchmount Park,Ionview,Kennedy Park
M1L       Scarborough                    Clairlea,Golden Mile,Oakridge
M1M       Scarborough    Cliffcrest,Cliffside,Scarborough Village West
M1N       Scarborough                       Birch Cliff,Cliffside West
Name: Neighborhood, dtype: object

### zip_grp is a tuple. Convert it back into dataframe and reset the index

In [11]:
zip_grp=pd.DataFrame(zip_grp).reset_index()

In [12]:
zip_grp.head(10)

Postcode      Borough                                   Neighborhood
0      M1B  Scarborough                                  Rouge,Malvern
1      M1C  Scarborough           Highland Creek,Rouge Hill,Port Union
2      M1E  Scarborough                Guildwood,Morningside,West Hill
3      M1G  Scarborough                                         Woburn
4      M1H  Scarborough                                      Cedarbrae
5      M1J  Scarborough                            Scarborough Village
6      M1K  Scarborough      East Birchmount Park,Ionview,Kennedy Park
7      M1L  Scarborough                  Clairlea,Golden Mile,Oakridge
8      M1M  Scarborough  Cliffcrest,Cliffside,Scarborough Village West
9      M1N  Scarborough                     Birch Cliff,Cliffside West

In [13]:
zip_grp.shape

(103, 3)

### The dataframe contains 103 lines and 3 columns

### I couldn't import the package, so I upload the coordinate file to github as the open source file

In [14]:
Location = pd.read_csv('https://raw.githubusercontent.com/pankeji/Coursera_Capstone/master/Geospatial_Coordinates.csv')

In [15]:
Location.head()

Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476

### Use concat function to merge the two dataframe

In [16]:
New_zip=pd.concat([zip_grp,Location],axis=1)
New_zip.head()

Postcode      Borough                          Neighborhood Postal Code  \
0      M1B  Scarborough                         Rouge,Malvern         M1B   
1      M1C  Scarborough  Highland Creek,Rouge Hill,Port Union         M1C   
2      M1E  Scarborough       Guildwood,Morningside,West Hill         M1E   
3      M1G  Scarborough                                Woburn         M1G   
4      M1H  Scarborough                             Cedarbrae         M1H   

    Latitude  Longitude  
0  43.806686 -79.194353  
1  43.784535 -79.160497  
2  43.763573 -79.188711  
3  43.770992 -79.216917  
4  43.773136 -79.239476

### As we can see the two columns Postcode and Postal Code are identical. Let's remove the second one.

In [17]:
New_zip.drop(['Postal Code'], axis=1, inplace=True)

In [18]:
New_zip.head(10)

Postcode      Borough                                   Neighborhood  \
0      M1B  Scarborough                                  Rouge,Malvern   
1      M1C  Scarborough           Highland Creek,Rouge Hill,Port Union   
2      M1E  Scarborough                Guildwood,Morningside,West Hill   
3      M1G  Scarborough                                         Woburn   
4      M1H  Scarborough                                      Cedarbrae   
5      M1J  Scarborough                            Scarborough Village   
6      M1K  Scarborough      East Birchmount Park,Ionview,Kennedy Park   
7      M1L  Scarborough                  Clairlea,Golden Mile,Oakridge   
8      M1M  Scarborough  Cliffcrest,Cliffside,Scarborough Village West   
9      M1N  Scarborough                     Birch Cliff,Cliffside West   

    Latitude  Longitude  
0  43.806686 -79.194353  
1  43.784535 -79.160497  
2  43.763573 -79.188711  
3  43.770992 -79.216917  
4  43.773136 -79.239476  
5  43.744734 -79.239476  
6  43.727929 -79.262029  
7  43.711112 -79.284577  
8  43.716316 -79.239476  
9  43.692657 -79.264848

### Let's count how many neighborhoods are in the Borough containing "Toronto"

In [19]:
New_zip['Borough'].value_counts()

North York          24
Downtown Toronto    19
Scarborough         17
Etobicoke           11
Central Toronto      9
West Toronto         6
East Toronto         5
East York            5
York                 5
Queen's Park         1
Mississauga          1
Name: Borough, dtype: int64

### we can see there are Downtown Toronto, Central Toronto, West Toronto, and East Toronto.

In [20]:
Toronto = New_zip[New_zip['Borough'].str.contains('Toronto')]
Toronto.head(20)

Postcode           Borough  \
37      M4E      East Toronto   
41      M4K      East Toronto   
42      M4L      East Toronto   
43      M4M      East Toronto   
44      M4N   Central Toronto   
45      M4P   Central Toronto   
46      M4R   Central Toronto   
47      M4S   Central Toronto   
48      M4T   Central Toronto   
49      M4V   Central Toronto   
50      M4W  Downtown Toronto   
51      M4X  Downtown Toronto   
52      M4Y  Downtown Toronto   
53      M5A  Downtown Toronto   
54      M5B  Downtown Toronto   
55      M5C  Downtown Toronto   
56      M5E  Downtown Toronto   
57      M5G  Downtown Toronto   
58      M5H  Downtown Toronto   
59      M5J  Downtown Toronto   

                                         Neighborhood   Latitude  Longitude  
37                                        The Beaches  43.676357 -79.293031  
41                        The Danforth West,Riverdale  43.679557 -79.352188  
42                      The Beaches West,India Bazaar  43.668999 -79.315572  
43                                    Studio District  43.659526 -79.340923  
44                                      Lawrence Park  43.728020 -79.388790  
45                                   Davisville North  43.712751 -79.390197  
46                                 North Toronto West  43.715383 -79.405678  
47                                         Davisville  43.704324 -79.388790  
48                         Moore Park,Summerhill East  43.689574 -79.383160  
49  Deer Park,Forest Hill SE,Rathnelly,South Hill,...  43.686412 -79.400049  
50                                           Rosedale  43.679563 -79.377529  
51                         Cabbagetown,St. James Town  43.667967 -79.367675  
52                               Church and Wellesley  43.665860 -79.383160  
53                                       Harbourfront  43.654260 -79.360636  
54                            Ryerson,Garden District  43.657162 -79.378937  
55                                     St. James Town  43.651494 -79.375418  
56                                        Berczy Park  43.644771 -79.373306  
57                                 Central Bay Street  43.657952 -79.387383  
58                             Adelaide,King,Richmond  43.650571 -79.384568  
59    Harbourfront East,Toronto Islands,Union Station  43.640816 -79.381752

In [21]:
map_Toronto = folium.Map(location=[43.6532, -79.3832], zoom_start=12)

for lat, lng, label in zip(Toronto['Latitude'], Toronto['Longitude'], Toronto['Neighborhood']):
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_Toronto)  
    
map_Toronto

In [22]:
CLIENT_ID = 'YPIH1XYOLTI4LWYX0VO1N3NHSANMNVWHYA52ODPB4PGOKXJI' # your Foursquare ID
CLIENT_SECRET = 'PG1W4GDQXPZVKVHQK1A1YATOTEOR4FL131JKX2EOC1OWIQSW' # your Foursquare Secret
VERSION = '20180605' # Foursquare API version

LIMIT=100
Radius=500

In [23]:
def getNearbyVenues(names, latitudes, longitudes, radius=Radius):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [24]:
Toronto_venues = getNearbyVenues(names=Toronto['Neighborhood'],
                                   latitudes=Toronto['Latitude'],
                                   longitudes=Toronto['Longitude']
                                  )

The Beaches
The Danforth West,Riverdale
The Beaches West,India Bazaar
Studio District
Lawrence Park
Davisville North
North Toronto West
Davisville
Moore Park,Summerhill East
Deer Park,Forest Hill SE,Rathnelly,South Hill,Summerhill West
Rosedale
Cabbagetown,St. James Town
Church and Wellesley
Harbourfront
Ryerson,Garden District
St. James Town
Berczy Park
Central Bay Street
Adelaide,King,Richmond
Harbourfront East,Toronto Islands,Union Station
Design Exchange,Toronto Dominion Centre
Commerce Court,Victoria Hotel
Roselawn
Forest Hill North,Forest Hill West
The Annex,North Midtown,Yorkville
Harbord,University of Toronto
Chinatown,Grange Park,Kensington Market
CN Tower,Bathurst Quay,Island airport,Harbourfront West,King and Spadina,Railway Lands,South Niagara
Stn A PO Boxes 25 The Esplanade
First Canadian Place,Underground city
Christie
Dovercourt Village,Dufferin
Little Portugal,Trinity
Brockton,Exhibition Place,Parkdale Village
High Park,The Junction South
Parkdale,Roncesvalles
Runnymede

In [25]:
print(Toronto_venues.shape)
Toronto_venues.head(10)

(1716, 7)


Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0                  The Beaches              43.676357              -79.293031   
1                  The Beaches              43.676357              -79.293031   
2                  The Beaches              43.676357              -79.293031   
3                  The Beaches              43.676357              -79.293031   
4  The Danforth West,Riverdale              43.679557              -79.352188   
5  The Danforth West,Riverdale              43.679557              -79.352188   
6  The Danforth West,Riverdale              43.679557              -79.352188   
7  The Danforth West,Riverdale              43.679557              -79.352188   
8  The Danforth West,Riverdale              43.679557              -79.352188   
9  The Danforth West,Riverdale              43.679557              -79.352188   

                                Venue  Venue Latitude  Venue Longitude  \
0                   Glen Manor Ravine       43.676821       -79.293942   
1  The Big Carrot Natural Food Market       43.678879       -79.297734   
2                 Grover Pub and Grub       43.679181       -79.297215   
3                       Upper Beaches       43.680563       -79.292869   
4                            Pantheon       43.677621       -79.351434   
5                       MenEssentials       43.677820       -79.351265   
6                     Cafe Fiorentina       43.677743       -79.350115   
7                        Dolce Gelato       43.677773       -79.351187   
8                               Mezes       43.677962       -79.350196   
9              Louis Cifer Brew Works       43.677663       -79.351313   

       Venue Category  
0               Trail  
1   Health Food Store  
2                 Pub  
3        Neighborhood  
4    Greek Restaurant  
5      Cosmetics Shop  
6  Italian Restaurant  
7      Ice Cream Shop  
8    Greek Restaurant  
9             Brewery

In [26]:
Toronto_venues['Neighborhood'].count()

1716

In [27]:
Toronto_onehot = pd.get_dummies(Toronto_venues[['Venue Category']], prefix="", prefix_sep="")
Toronto_onehot.shape

(1716, 239)

In [28]:
Toronto_onehot['Neighborhood'].head()

0    0
1    0
2    0
3    1
4    0
Name: Neighborhood, dtype: uint8

### As we can see, when we used get_dummies, a new column with the name of "Neighborhood" appeared. It is because in the category column, there is a feature called "Neighborhood". In order to add the real Neighborhood name back into the data, we need to rename the original "Neighbohood" column with a new name such as "Neighborhood_name".

In [29]:
Toronto_Neigh=pd.DataFrame(Toronto_venues['Neighborhood'])
Toronto_Neigh.rename(columns={'Neighborhood':'Neighborhood_name'}, inplace=True)
Toronto_Neigh.head()


Neighborhood_name
0                  The Beaches
1                  The Beaches
2                  The Beaches
3                  The Beaches
4  The Danforth West,Riverdale

In [30]:
Toronto_onehot=pd.concat([Toronto_Neigh,Toronto_onehot], axis=1 )
Toronto_onehot.head()

Neighborhood_name  Afghan Restaurant  Airport  \
0                  The Beaches                  0        0   
1                  The Beaches                  0        0   
2                  The Beaches                  0        0   
3                  The Beaches                  0        0   
4  The Danforth West,Riverdale                  0        0   

   Airport Food Court  Airport Gate  Airport Lounge  Airport Service  \
0                   0             0               0                0   
1                   0             0               0                0   
2                   0             0               0                0   
3                   0             0               0                0   
4                   0             0               0                0   

   Airport Terminal  American Restaurant  Antique Shop  ...  Toy / Game Store  \
0                 0                    0             0  ...                 0   
1                 0                    0             0  ...                 0   
2                 0                    0             0  ...                 0   
3                 0                    0             0  ...                 0   
4                 0                    0             0  ...                 0   

   Trail  Train Station  Vegetarian / Vegan Restaurant  Video Game Store  \
0      1              0                              0                 0   
1      0              0                              0                 0   
2      0              0                              0                 0   
3      0              0                              0                 0   
4      0              0                              0                 0   

   Vietnamese Restaurant  Wine Bar  Wine Shop  Women's Store  Yoga Studio  
0                      0         0          0              0            0  
1                      0         0          0              0            0  
2                      0         0          0              0            0  
3                      0         0          0              0            0  
4                      0         0          0              0            0  

[5 rows x 240 columns]

In [31]:
Toronto_onehot['Neighborhood_name'].head(10)

0                    The Beaches
1                    The Beaches
2                    The Beaches
3                    The Beaches
4    The Danforth West,Riverdale
5    The Danforth West,Riverdale
6    The Danforth West,Riverdale
7    The Danforth West,Riverdale
8    The Danforth West,Riverdale
9    The Danforth West,Riverdale
Name: Neighborhood_name, dtype: object

### Update the feature requency

In [32]:
Toronto_grp = Toronto_onehot.groupby('Neighborhood_name').mean().reset_index()
Toronto_grp

Neighborhood_name  Afghan Restaurant  \
0                              Adelaide,King,Richmond           0.000000   
1                                         Berczy Park           0.000000   
2          Brockton,Exhibition Place,Parkdale Village           0.000000   
3   Business Reply Mail Processing Centre 969 Eastern           0.000000   
4   CN Tower,Bathurst Quay,Island airport,Harbourf...           0.000000   
5                          Cabbagetown,St. James Town           0.000000   
6                                  Central Bay Street           0.000000   
7             Chinatown,Grange Park,Kensington Market           0.000000   
8                                            Christie           0.000000   
9                                Church and Wellesley           0.012346   
10                      Commerce Court,Victoria Hotel           0.000000   
11                                         Davisville           0.000000   
12                                   Davisville North           0.000000   
13  Deer Park,Forest Hill SE,Rathnelly,South Hill,...           0.000000   
14            Design Exchange,Toronto Dominion Centre           0.000000   
15                        Dovercourt Village,Dufferin           0.000000   
16              First Canadian Place,Underground city           0.000000   
17                 Forest Hill North,Forest Hill West           0.000000   
18                      Harbord,University of Toronto           0.000000   
19                                       Harbourfront           0.000000   
20    Harbourfront East,Toronto Islands,Union Station           0.000000   
21                       High Park,The Junction South           0.000000   
22                                      Lawrence Park           0.000000   
23                            Little Portugal,Trinity           0.000000   
24                         Moore Park,Summerhill East           0.000000   
25                                 North Toronto West           0.000000   
26                              Parkdale,Roncesvalles           0.000000   
27                                       Queen's Park           0.000000   
28                                           Rosedale           0.000000   
29                                           Roselawn           0.000000   
30                                  Runnymede,Swansea           0.000000   
31                            Ryerson,Garden District           0.000000   
32                                     St. James Town           0.000000   
33                    Stn A PO Boxes 25 The Esplanade           0.000000   
34                                    Studio District           0.000000   
35                  The Annex,North Midtown,Yorkville           0.000000   
36                                        The Beaches           0.000000   
37                      The Beaches West,India Bazaar           0.000000   
38                        The Danforth West,Riverdale           0.000000   

     Airport  Airport Food Court  Airport Gate  Airport Lounge  \
0   0.000000            0.000000      0.000000        0.000000   
1   0.000000            0.000000      0.000000        0.000000   
2   0.000000            0.000000      0.000000        0.000000   
3   0.000000            0.000000      0.000000        0.000000   
4   0.058824            0.058824      0.058824        0.117647   
5   0.000000            0.000000      0.000000        0.000000   
6   0.000000            0.000000      0.000000        0.000000   
7   0.000000            0.000000      0.000000        0.000000   
8   0.000000            0.000000      0.000000        0.000000   
9   0.000000            0.000000      0.000000        0.000000   
10  0.000000            0.000000      0.000000        0.000000   
11  0.000000            0.000000      0.000000        0.000000   
12  0.000000            0.000000      0.000000        0.000000   
13  0.000000            0.000000      0.000000        0.000000   
14  0

### Create a new dataframe containing top 10 venues in each neighborhood

In [33]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [34]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = Toronto_grp['Neighborhood_name']

for ind in np.arange(Toronto_grp.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(Toronto_grp.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood 1st Most Common Venue  \
0                             Adelaide,King,Richmond           Coffee Shop   
1                                        Berczy Park           Coffee Shop   
2         Brockton,Exhibition Place,Parkdale Village                  Café   
3  Business Reply Mail Processing Centre 969 Eastern            Smoke Shop   
4  CN Tower,Bathurst Quay,Island airport,Harbourf...       Airport Service   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0            Steakhouse                   Bar                  Café   
1          Cocktail Bar              Beer Bar                Bakery   
2           Coffee Shop        Breakfast Spot           Yoga Studio   
3         Auto Workshop                  Park           Pizza Place   
4        Airport Lounge      Airport Terminal       Harbor / Marina   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0                 Hotel      Sushi Restaurant       Thai Restaurant   
1            Steakhouse    Seafood Restaurant           Cheese Shop   
2               Stadium         Burrito Place            Restaurant   
3      Recording Studio            Restaurant         Burrito Place   
4                 Plane              Boutique         Boat or Ferry   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0          Burger Joint      Asian Restaurant         Clothing Store  
1                  Café        Farmers Market                   Park  
2          Climbing Gym         Grocery Store                 Bakery  
3               Brewery            Skate Park     Light Rail Station  
4                   Bar           Coffee Shop       Sculpture Garden

Toronto_grp.drop('Neighborhood_name', 1).head()

### Use K-means method. After several trying, cluster number was set at 7 which can provide a more meaningful analysis

In [35]:
# set number of clusters
kclusters = 7

Toronto_clustering = Toronto_grp.drop('Neighborhood_name', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(Toronto_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 
pd.DataFrame(kmeans.cluster_centers_)

0             1             2             3             4         5    \
0  0.000386 -8.673617e-19 -8.673617e-19 -8.673617e-19 -1.734723e-18  0.000000   
1  0.000000  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00  0.000000   
2  0.000000  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00  0.000000   
3  0.000000  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00  0.000000   
4  0.000000  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00  0.000000   
5  0.000000  5.882353e-02  5.882353e-02  5.882353e-02  1.176471e-01  0.176471   
6  0.000000  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00  0.000000   

            6         7         8         9    ...       229       230  \
0 -1.734723e-18  0.010954  0.000994  0.001563  ...  0.002534  0.000744   
1  0.000000e+00  0.000000  0.000000  0.000000  ...  0.000000  0.000000   
2  0.000000e+00  0.000000  0.000000  0.000000  ...  0.000000  0.250000   
3  0.000000e+00  0.000000  0.000000  0.000000  ...  0.000000  0.100000   
4  0.000000e+00  0.000000  0.000000  0.000000  ...  0.000000  0.250000   
5  1.176471e-01  0.000000  0.000000  0.000000  ...  0.000000  0.000000   
6  0.000000e+00  0.000000  0.000000  0.000000  ...  0.000000  0.000000   

        231       232       233       234       235       236       237  \
0  0.000938  0.008674  0.001181  0.005654  0.003594  0.002224  0.000625   
1  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000   
2  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000   
3  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000   
4  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000   
5  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000   
6  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000   

        238  
0  0.007186  
1  0.000000  
2  0.000000  
3  0.000000  
4  0.000000  
5  0.000000  
6  0.000000  

[7 rows x 239 columns]

In [36]:
# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

Toronto_merged = Toronto

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
Toronto_merged = Toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')

Toronto_merged.head() # check the last columns!

Postcode          Borough                   Neighborhood   Latitude  \
37      M4E     East Toronto                    The Beaches  43.676357   
41      M4K     East Toronto    The Danforth West,Riverdale  43.679557   
42      M4L     East Toronto  The Beaches West,India Bazaar  43.668999   
43      M4M     East Toronto                Studio District  43.659526   
44      M4N  Central Toronto                  Lawrence Park  43.728020   

    Longitude  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
37 -79.293031               4                 Trail     Health Food Store   
41 -79.352188               0      Greek Restaurant           Coffee Shop   
42 -79.315572               0        Sandwich Place                  Park   
43 -79.340923               0                  Café           Coffee Shop   
44 -79.388790               3  Gym / Fitness Center                  Park   

   3rd Most Common Venue 4th Most Common Venue        5th Most Common Venue  \
37          Neighborhood                   Pub  Eastern European Restaurant   
41    Italian Restaurant        Ice Cream Shop       Furniture / Home Store   
42             Pet Store                   Pub                 Burger Joint   
43                Bakery    Italian Restaurant                    Gastropub   
44              Bus Line           Swim School               Discount Store   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
37               Dog Run      Doner Restaurant            Donut Shop   
41           Yoga Studio               Brewery            Sports Bar   
42         Burrito Place  Fast Food Restaurant     Fish & Chips Shop   
43   American Restaurant               Brewery           Yoga Studio   
44        Farmers Market    Falafel Restaurant           Event Space   

        9th Most Common Venue 10th Most Common Venue  
37        Dumpling Restaurant            Yoga Studio  
41                        Spa         Cosmetics Shop  
42         Italian Restaurant             Steakhouse  
43  Latin American Restaurant            Fish Market  
44       Ethiopian Restaurant    Empanada Restaurant

### Generate the cluster map for Toronto

In [37]:
# create map
map_clusters = folium.Map(location=[43.6532, -79.3832], zoom_start=12)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(Toronto_merged['Latitude'], Toronto_merged['Longitude'], Toronto_merged['Neighborhood'], Toronto_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

## Cluster 1

In [38]:
Toronto_merged.loc[Toronto_merged['Cluster Labels'] == 0, Toronto_merged.columns[[1] + list(range(5, Toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
41      East Toronto               0      Greek Restaurant   
42      East Toronto               0        Sandwich Place   
43      East Toronto               0                  Café   
45   Central Toronto               0                   Gym   
46   Central Toronto               0        Clothing Store   
47   Central Toronto               0          Dessert Shop   
49   Central Toronto               0                   Pub   
51  Downtown Toronto               0           Coffee Shop   
52  Downtown Toronto               0           Coffee Shop   
53  Downtown Toronto               0           Coffee Shop   
54  Downtown Toronto               0           Coffee Shop   
55  Downtown Toronto               0           Coffee Shop   
56  Downtown Toronto               0           Coffee Shop   
57  Downtown Toronto               0           Coffee Shop   
58  Downtown Toronto               0           Coffee Shop   
59  Downtown Toronto               0           Coffee Shop   
60  Downtown Toronto               0           Coffee Shop   
61  Downtown Toronto               0           Coffee Shop   
65   Central Toronto               0                  Café   
66  Downtown Toronto               0                  Café   
67  Downtown Toronto               0                  Café   
69  Downtown Toronto               0           Coffee Shop   
70  Downtown Toronto               0           Coffee Shop   
75  Downtown Toronto               0         Grocery Store   
76      West Toronto               0                Bakery   
77      West Toronto               0                   Bar   
78      West Toronto               0                  Café   
82      West Toronto               0       Thai Restaurant   
83      West Toronto               0             Gift Shop   
84      West Toronto               0           Coffee Shop   
85  Downtown Toronto               0           Coffee Shop   
87      East Toronto               0            Smoke Shop   

    2nd Most Common Venue          3rd Most Common Venue  \
41            Coffee Shop             Italian Restaurant   
42                   Park                      Pet Store   
43            Coffee Shop                         Bakery   
45           Dance Studio                 Sandwich Place   
46            Coffee Shop            Sporting Goods Shop   
47         Sandwich Place                    Pizza Place   
49            Coffee Shop            American Restaurant   
51             Restaurant                           Café   
52       Sushi Restaurant            Japanese Restaurant   
53                 Bakery                           Park   
54         Clothing Store                           Café   
55                   Café                     Restaurant   
56           Cocktail Bar                       Beer Bar   
57                   Café             Italian Restaurant   
58             Steakhouse                            Bar   
59               Aquarium                           Café   
60                  Hotel                           Café   
61                   Café                          Hotel   
65         Sandwich Place                    Coffee Shop   
66             Restaurant                            Bar   
67  Vietnamese Restaurant  Vegetarian / Vegan Restaurant   
69                   Café             Seafood Restaurant   
70                   Café                     Steakhouse   
75                   Café                           Park   
76               Pharmacy                  Grocery Store   
77       Asian Restaurant                     Restaurant   
78            Coffee Shop                 Breakfast Spot   
82                   Café             Mexican Restaurant   
83              Bookstore                   Dessert Shop   
84                   Café               Sushi Restaurant   
85           Burger Joint                            Gym   
87          Auto Workshop                     

## Cluster 2

In [39]:
Toronto_merged.loc[Toronto_merged['Cluster Labels'] == 1, Toronto_merged.columns[[1] + list(range(5, Toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
48  Central Toronto               1                   Gym   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
48            Playground            Restaurant            Comic Shop   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
48                 Diner    Falafel Restaurant           Event Space   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
48  Ethiopian Restaurant   Empanada Restaurant      Electronics Store

## Cluster 3

In [40]:
Toronto_merged.loc[Toronto_merged['Cluster Labels'] == 2, Toronto_merged.columns[[1] + list(range(5, Toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
50  Downtown Toronto               2                  Park   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
50                 Trail            Playground     Electronics Store   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
50               Dog Run      Doner Restaurant            Donut Shop   

   8th Most Common Venue        9th Most Common Venue 10th Most Common Venue  
50   Dumpling Restaurant  Eastern European Restaurant            Yoga Studio

## Cluster 4

In [41]:
Toronto_merged.loc[Toronto_merged['Cluster Labels'] == 3, Toronto_merged.columns[[1] + list(range(5, Toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
44  Central Toronto               3  Gym / Fitness Center   
64  Central Toronto               3                 Trail   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
44                  Park              Bus Line           Swim School   
64                  Park      Sushi Restaurant              Bus Line   

   5th Most Common Venue        6th Most Common Venue 7th Most Common Venue  \
44        Discount Store               Farmers Market    Falafel Restaurant   
64         Jewelry Store  Eastern European Restaurant      Doner Restaurant   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
44           Event Space  Ethiopian Restaurant    Empanada Restaurant  
64            Donut Shop   Dumpling Restaurant            Yoga Studio

## Cluster 5

In [42]:
Toronto_merged.loc[Toronto_merged['Cluster Labels'] == 4, Toronto_merged.columns[[1] + list(range(5, Toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
37  East Toronto               4                 Trail     Health Food Store   

   3rd Most Common Venue 4th Most Common Venue        5th Most Common Venue  \
37          Neighborhood                   Pub  Eastern European Restaurant   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
37               Dog Run      Doner Restaurant            Donut Shop   

   9th Most Common Venue 10th Most Common Venue  
37   Dumpling Restaurant            Yoga Studio

## Cluster 6

In [43]:
Toronto_merged.loc[Toronto_merged['Cluster Labels'] == 5, Toronto_merged.columns[[1] + list(range(5, Toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
68  Downtown Toronto               5       Airport Service   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
68        Airport Lounge      Airport Terminal       Harbor / Marina   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
68                 Plane              Boutique         Boat or Ferry   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
68                   Bar           Coffee Shop       Sculpture Garden

## Cluster 7

In [44]:
Toronto_merged.loc[Toronto_merged['Cluster Labels'] == 6, Toronto_merged.columns[[1] + list(range(5, Toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
63  Central Toronto               6                Garden   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
63                  Pool          Home Service           Yoga Studio   

          5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
63  Eastern European Restaurant               Dog Run      Doner Restaurant   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
63            Donut Shop   Dumpling Restaurant    Empanada Restaurant